clang/test/CodeGen/cheri/cheri-union-i128.c

```diff
@@ -0,0 +1,132 @@
+// CHECK-32-NEXT:  entry:
+// CHECK-32-NEXT:    [[U:%.*]] = alloca [[UNION_U64:%.*]], align 8, addrspace(200)
+// CHECK-32-NEXT:    [[COERCE_DIVE:%.*]] = getelementptr inbounds [[UNION_U64]], ptr addrspace(200) [[U]], i32 0, i32 0
+// CHECK-32-NEXT:    store i64 [[U_COERCE]], ptr addrspace(200) [[COERCE_DIVE]], align 8
+// CHECK-32-NEXT:    [[TMP0:%.*]] = load i64, ptr addrspace(200) [[U]], align 8
+// CHECK-32-NEXT:    ret i64 [[TMP0]]
+//
+// CHECK-64-LABEL: define {{[^@]+}}@arg64
+// CHECK-64-SAME: (ptr addrspace(200) [[U_COERCE:%.*]]) addrspace(200) #[[ATTR0:[0-9]+]] {
+// CHECK-64-NEXT:  entry:
+// CHECK-64-NEXT:    [[U:%.*]] = alloca [[UNION_U64:%.*]], align 16, addrspace(200)
+// CHECK-64-NEXT:    [[COERCE_DIVE:%.*]] = getelementptr inbounds [[UNION_U64]], ptr addrspace(200) [[U]], i32 0, i32 0
+// CHECK-64-NEXT:    store ptr addrspace(200) [[U_COERCE]], ptr addrspace(200) [[COERCE_DIVE]], align 16
+// CHECK-64-NEXT:    [[TMP0:%.*]] = load i64, ptr addrspace(200) [[U]], align 16
+// CHECK-64-NEXT:    ret i64 [[TMP0]]
+//
+__INT64_TYPE__ arg64(union u64 u) {
+  return u.x;
+}
+// CHECK-32-LABEL: define {{[^@]+}}@call64
+// CHECK-32-SAME: () addrspace(200) #[[ATTR0]] {
+// CHECK-32-NEXT:  entry:
+// CHECK-32-NEXT:    [[TMP0:%.*]] = load i64, ptr addrspace(200) @global64, align 8
+// CHECK-32-NEXT:    [[CALL:%.*]] = call i64 @arg64(i64 [[TMP0]])
+// CHECK-32-NEXT:    ret void
+//
+// CHECK-64-LABEL: define {{[^@]+}}@call64
+// CHECK-64-SAME: () addrspace(200) #[[ATTR0]] {
+// CHECK-64-NEXT:  entry:
+// CHECK-64-NEXT:    [[TMP0:%.*]] = load ptr addrspace(200), ptr addrspace(200) @global64, align 16
+// CHECK-64-NEXT:    [[CALL:%.*]] = call i64 @arg64(ptr addrspace(200) [[TMP0]])
+// CHECK-64-NEXT:    ret void
+//
+void call64(void) {
+  arg64(global64);
+}
+
+// CHECK-32-LABEL: define {{[^@]+}}@ret64
+// CHECK-32-SAME: () addrspace(200) #[[ATTR0]] {
+// CHECK-32-NEXT:  entry:
+// CHECK-32-NEXT:    [[RETVAL:%.*]] = alloca [[UNION_U64:%.*]], align 8, addrspace(200)
+// CHECK-32-NEXT:    call void @llvm.memcpy.p200.p200.i32(ptr addrspace(200) align 8 [[RETVAL]], ptr addrspace(200) align 8 @global64, i32 8, i1 false)
+// CHECK-32-NEXT:    [[TMP0:%.*]] = load [[UNION_U64]], ptr addrspace(200) [[RETVAL]], align 8
+// CHECK-32-NEXT:    ret [[UNION_U64]] [[TMP0]]
+//
+// CHECK-64-LABEL: define {{[^@]+}}@ret64
+// CHECK-64-SAME: () addrspace(200) #[[ATTR0]] {
+// CHECK-64-NEXT:  entry:
+// CHECK-64-NEXT:    [[RETVAL:%.*]] = alloca [[UNION_U64:%.*]], align 16, addrspace(200)
+// CHECK-64-NEXT:    call void @llvm.memcpy.p200.p200.i64(ptr addrspace(200) align 16 [[RETVAL]], ptr addrspace(200) align 16 @global64, i64 16, i1 false)
+// CHECK-64-NEXT:    [[TMP0:%.*]] = load [[UNION_U64]], ptr addrspace(200) [[RETVAL]], align 16
+// CHECK-64-NEXT:    ret [[UNION_U64]] [[TMP0]]
+//
+union u64 ret64(void) {
+  return global64;
+}
+
+// FIXME: Unions with int128+capability members should not be passed as i128!
+union u128 {
+  __int128 x;
```

Comment:
Huh, I thought 32-bit didn't support __int128 in Clang...

Reply:
I thought so too, but apparently it works with -cc1 command lines...

Reply:
Aha turns out we have a hack in getPointerWidth(0), so for RV32 we report true for hasint128:
`return (getPointerWidth(0) >= 64) || getTargetOpts().ForceEnableInt128;`

Reply:
I've added `-fforce-enable-int128` to keep the test working once that has been fixed.